jmx-metrics/src/main/groovy/io/opentelemetry/contrib/jmxmetrics/GroovyMetricEnvironment.java

```diff
@@ -331,55 +471,99 @@ public void registerDoubleUpDownCounterCallback(
+        ObservableLongMeasurement finalObs = (ObservableLongMeasurement) obs;
+        meter.batchCallback(() -> cb.accept(finalObs), obs);
+      }
+    } else if (updater != null) {
+      // If the instrument has already been built with the appropriate proxied observer,
+      // update the registry so that the callback has the appropriate updater function
+      AtomicReference<Consumer<ObservableLongMeasurement>> existingUpdater =
+          longUpdaterRegistry.get(descriptorHash);
+      existingUpdater.set(updater);
+    }
+
+    return (ObservableLongMeasurement) obs;
+  }
+
+  /**
+   * Register a collection of observables in a single batch callback
+   *
+   * @param identifier - object used to identify the callback to have only one callback
+   * @param callback - closure that records measurements for the observables
+   * @param measurement - first observable, the SDK expects this is always collected
+   * @param additional - remaining observable, the SDK expects this is sometimes collected
+   */
+  public void registerBatchCallback(
+      Object identifier,
+      Closure<?> callback,
+      ObservableMeasurement measurement,
+      ObservableMeasurement... additional) {
+    int hash = identifier.hashCode();
+    // Store the callback in the registry so the proxied callback always runs the latest
+    // metric collection closure
+    AtomicReference<Closure<?>> existing =
+        batchUpdaterRegistry.putIfAbsent(hash, new AtomicReference<>());
+    batchUpdaterRegistry.get(hash).set(callback);
+
+    // If the callback had previously been stored, we assume the back callback is already
+    // registered with the Meter
+    if (existing == null) {
+      meter.batchCallback(
+          () -> batchUpdaterRegistry.get(hash).get().call(), measurement, additional);
+    }
   }
 
   private Consumer<ObservableDoubleMeasurement> proxiedDoubleObserver(
-      final String name,
-      final String description,
-      final String unit,
-      final InstrumentType instrumentType,
-      final Consumer<ObservableDoubleMeasurement> updater) {
-    InstrumentDescriptor descriptor =
-        InstrumentDescriptor.create(
-            name, description, unit, instrumentType, InstrumentValueType.DOUBLE);
-    doubleUpdaterRegistry.putIfAbsent(descriptor.hashCode(), new AtomicReference<>());
+      final int descriptorHash, final Consumer<ObservableDoubleMeasurement> updater) {
+    doubleUpdaterRegistry.putIfAbsent(descriptorHash, new AtomicReference<>());
     AtomicReference<Consumer<ObservableDoubleMeasurement>> existingUpdater =
-        doubleUpdaterRegistry.get(descriptor.hashCode());
+        doubleUpdaterRegistry.get(descriptorHash);
     existingUpdater.set(updater);
     return doubleResult -> {
-      Consumer<ObservableDoubleMeasurement> existing = existingUpdater.get();
+      Consumer<ObservableDoubleMeasurement> existing =
+          doubleUpdaterRegistry.get(descriptorHash).get();
```

Comment:
Is this indirection necessary?

Reply:
Storing it into the local variable before calling accept? No, I just didn't refactor that from the existing code. Will do.